compiler/rustc_builtin_macros/src/assert/context.rs

```diff
@@ -0,0 +1,434 @@
+use rustc_ast::{
+    ptr::P,
+    token,
+    token::Nonterminal,
+    tokenstream::CanSynthesizeMissingTokens,
+    tokenstream::{DelimSpan, TokenStream, TokenTree},
+    Block, Expr, ExprKind, ItemKind, LitKind, MacArgs, MacCall, MacDelimiter, Path, PathSegment,
+    Stmt, StmtKind, StrStyle, StructRest, UseTree, UseTreeKind, DUMMY_NODE_ID,
+};
+use rustc_ast_pretty::pprust;
+use rustc_expand::base::ExtCtxt;
+use rustc_span::{
+    symbol::{sym, Ident, Symbol},
+    Span,
+};
+
+pub(super) struct Context<'cx, 'a> {
+    pub(super) cx: &'cx ExtCtxt<'a>,
+    pub(super) span: Span,
+    captures: Vec<Capture>,
+}
+
+impl<'cx, 'a> Context<'cx, 'a> {
+    pub(super) fn new(cx: &'cx ExtCtxt<'a>, span: Span) -> Self {
+        Self { cx, captures: Vec::new(), span }
+    }
+
+    /// Builds the whole `assert!` expression.
+    ///
+    /// {
+    ///    use ::core::asserting::*;
+    ///
+    ///    let mut __capture0 = Capture::Unevaluated;
+    ///    ...
+    ///
+    ///    if !{ ... } {
+    ///        panic!(
+    ///            "Assertion failed: ... \n With expansion: ...",
+    ///            (&__capture0).printer_kind().new(" ... ", __capture0),
+    ///            ...
+    ///        );
+    ///    }
+    /// }
+    pub(super) fn build(mut self, mut cond_expr: P<Expr>, panic_path: Path) -> P<Expr> {
+        let expr_str = escape_to_fmt(&pprust::expr_to_string(&cond_expr));
+        let fmt_str = &pprust::expr_to_string_cb(&cond_expr, |expr| {
+            if let ExprKind::Path(..) = expr.kind { Some("{:?}") } else { None }
```

Comment:
How deep do you plan to detect paths in the condition expression?
Will paths will be replaced with `{:?}`s even in items inside blocks with this approach?

If yes, then it's incorrect.
If no, then we are pulling assert-specific logic to the pretty-printer, which is also bad.

One alternative is to take a mutable AST visitor (`MutVisitor`) and replace all paths in `cond_expr` with some unique placeholder identifier up to the necessary depth (the identifier must be illegal in Rust lexical grammar though, otherwise it could occur in `cond_expr` naturally).
Then placeholders can be replaced with `{:?}`s by string replacement in the result of `pprust::expr_to_string`.

Also, what if `{:?}` occurs naturally in `cond_expr`? Then it should be escaped somehow.

Reply:
Before jumping into it, it is worth noting that the RFC stated that the implementation is free to choose the showed message. That said, it is possible to do something like:

```
Assertion failed: a == 3 && b == 5 && c == 4
With captures:
    a = 1
    b = 5
    c = 4
```

And avoid this whole formatting complexity. What do you think about it?

Reply:
The RFC was originally accepted as a high-level goal of making the output of `assert` prettier, not as some specific output like described in the RFC.
So it should be entirely ok.
Also it's not necessary to implement all the niceties in one go.

Reply:
Thanks @petrochenkov 

I will proceed with `With captures ...` at least for now